src/renderers/shared/stack/reconciler/ReactCompositeComponent.js

```diff
@@ -833,11 +835,27 @@ var ReactCompositeComponentMixin = {
       this._currentElement = nextParentElement;
       this._context = nextUnmaskedContext;
       inst.props = nextProps;
-      inst.state = nextState;
+      this._updateState(nextState);
       inst.context = nextContext;
     }
   },
 
+  _updateState(nextState) {
```

Comment:
@gaearon I had to extract _notifyStateChanged again, so it can be used for notifying on inital state.